threadpoolctl.py

```diff
@@ -383,6 +397,36 @@ def _get_current_backend(self):
         get_backend_(backend, ctypes.sizeof(backend))
         return backend.value.decode("utf-8")
 
+    def switch_backend(self, backend):
+        """Switch the backend of FlexiBLAS
+
+        Parameters
+        ----------
+        backend : str
+            The name or the path to the shared library of the backend to switch to. If
+            the backend is not already loaded, it will be loaded first.
+        """
+        if backend not in self.loaded_backends:
+            if backend in self.available_backends:
+                load_func = getattr(self.dynlib, "flexiblas_load_backend", lambda _: -1)
+            else:  # assume backend is a path to a shared library
+                load_func = getattr(
+                    self.dynlib, "flexiblas_load_backend_library", lambda _: -1
+                )
+            res = load_func(str(backend).encode("utf-8"))
+            if res == -1:
+                raise RuntimeError(f"Failed to load backend {backend!r}.")
+
+            # Trigger a new search of loaded shared libraries since loading a new
+            # backend caused a dlopen.
+            self.parent._load_libraries()
```

Comment:
adding support to switch backends comes with a cost. When loading a new backend, we need to rerun the dl_iterate_phdr search for shared libraries.

We specifically designed the ThreadpoolController to avoid rerunning the search every time we call `info` or `limit`, for the reason that "it's unlikely that the loaded shared libraries will change during the lifetime of the program". This is no longer true with FlexiBLAS.

I think the most reasonable thing to do is to only trigger a new search when loading a new backend, even though it makes the LibControllers have a ref to the parent ThreadpoolController (which is the one able to rerun the search).

Reply:
> adding support to switch backends comes with a cost. When loading a new backend, we need to rerun the dl_iterate_phdr search for shared libraries.

I think this is fine as loading a new backend should not happen very often, or at least not in performance critical sections of a program.

> I think the most reasonable thing to do is to only trigger a new search when loading a new backend, even though it makes the LibControllers have a ref to the parent ThreadpoolController (which is the one able to rerun the search).

I agree, although I have not yet reviewed the diff yet.